src/jdk.jdwp.agent/share/native/libjdwp/threadControl.c

```diff
@@ -1720,12 +1727,70 @@ threadControl_isDebugThread(jthread thread)
     return rc;
 }
 
+/*** Start of support for StackFrame.PopFrames ***/
+
+/*
+ * Synchronization with popFrameEventLock and popFrameProceedLock
+ *
+ * StackFrame.PopFrames support has some complicated synchronization because
+ * JVMTI PopFrame() is called on the JDWP Command Reader thread (that's where
+ * the StackFrame.PopFrames command arrived on), but it needs to co-ordinate
+ * with the PopFrames target thread. The Command Reader thread has to call JVMTI
+ * PopFrame() on the target thread one frame at a time, resume the target thread,
+ * and then know when the target thread is done with the PopFrame() so it can be
+ * suspended so the next PopFrame() can then be done. Part of what makes this
+ * complex is that the target thread needs to have SINGLE_STEP enabled so we
+ * can tell when the PopFrame() has completed. (For reasons I don't understand,
+ * we don't rely on FRAME_POP events).
+ *
+ * Synchronization is done via the popFrameProceedLock and popFrameEventLock. Both
+ * the JDWP Command Reader thread and the PopFrame() target thread will grab
+ * both of these locks. However, one curious trait of these two locks is that
+ * the these two threads do not both grab them in the same order (and they need
+ * to do it this way). This seems like it should lead to a deadlock, but doesn't
+ * because of how each thread grabs these locks and waits or notifies on them.
+ *
+ * Command Reader Thread:
+ *  grab popFrameEventLock (which is always available)
+ *  for each frame
+ *    call PopFrame() on target thread
+ *    resume target thread
+ *    wait on popFrameEventLock (which releases it)
+ *    grab popFrameProceedLock
+ *    suspend target thread
+ *    notify on popFrameProceedLock
+ *    release popFrameProceedLock
+ *  }
+ *  release popFrameEventLock
+ *
+ * Target Thread:
+ *   for each frame pop (indicated by a SingleStep event)
+ *     grab popFrameProceedLock (which is always available)
+ *     grab popFrameEventLock (which Command Reader Thread is waiting on)
+ *     notify on popFrameEventLock
+ *     release popFrameEventLock
+ *     wait on popFrameProceedLock
+ *     release popFrameProceedLock
+ *   }
+ *
+ * The reason we don't get a deadlock is because the Reader thread always
+ * starts the process, so it always enters popFrameEventLock and then
+ * waits on it (which releases the lock). Therefore there is never a race for
+ * popFrameEventLock. Meanwhile when the target thread gets the SINGLE_STEP event,
+ * it always enters popFrameProceedLock first (which is always available), then
+ * popFrameProceedLock second. It will always succeed because the Reader thread
```

Comment:
> Ironically right now I'm looking at a very rare deadlock that involves this code. It doesn't seem to happen when I disabled ranked locking. It might be instigated by the dbgRawMonitor that ranked locking uses.

@kevinjwalls I tracked down this deadlock and filed [JDK-8332738](https://bugs.openjdk.org/browse/JDK-8332738). It's really a pre-existing issue, but we get lucky with the current implementation because RawMonitorExit does not self suspend until after releasing the monitor, thus avoiding the deadlock. The ranked monitors implementation adds a self suspend opportunity when we release a raw monitor, which is it hits this bug. This is a very ugly issue that involves two threads getting events at the same time, and the debugger doing a StackFrame.PopFrames. Fortunately the "pop frames" locks are not involved.

Reply:
I should also add that this issue was only turning up when I had virtual threads enabled, because that caused one of the threads to trigger a class load of some CarrierThread inner class the first time it parked. This generated a ClassPrepare event with unfortunately timing (it would be on Thread 2 of my description in [JDK-8332738](https://bugs.openjdk.org/browse/JDK-8332738)). However, I now have a much more direct test that doesn't require virtual threads and more readily reproduces the deadlock.